audit-data/2023-09-01-puppy-raffle.md

```diff
@@ -404,6 +405,134 @@ contract AttackerContract {
 
 **Recommended Mitigation:** Favor pull-payments over push-payments. This means modifying the `selectWinner` function so that the winner account has to claim the prize by calling a function, instead of having the contract automatically send the funds during execution of `selectWinner`.
 
+### [H-5] `PuppyRaffle::refund` does not update the length of the players array, which leads to a loss of fees and unability to pay out the winners
+
+**Description:** The `PuppyRaffle::refund` function does not update the length of the players array, which is used to determine the final prize pool at the `PuppyRaffle::selectWinner` function:
+```solidity
+    function refund(uint256 playerIndex) public {
+        address playerAddress = players[playerIndex];
+        require(playerAddress == msg.sender, "PuppyRaffle: Only the player can refund");
+        require(playerAddress != address(0), "PuppyRaffle: Player already refunded, or is not active");
+
+        payable(msg.sender).sendValue(entranceFee);
+
+->      players[playerIndex] = address(0);
+        emit RaffleRefunded(playerAddress);
+    }
+```
+We reset the player's address to `address(0)`, but we don't update the length of the players array, that's used in the `PuppyRaffle::selectWinner` function:
+```solidity
+    function selectWinner() external {
+        ...
+->      uint256 totalAmountCollected = players.length * entranceFee;
+        uint256 prizePool = (totalAmountCollected * 80) / 100;
+        uint256 fee = (totalAmountCollected * 20) / 100;
+        ...
+    }
+```
+We also don't check that the winner is not `address(0)` in the `PuppyRaffle::selectWinner` function. That means that a zero address could be selected as winner, and the contract would send the prize pool to the zero address.
+
+**Impact:** This could not only lead to a loss of fees (the contract will pay the winner more than it was supposed to) but also to a loss of the ability to pay out the winners, as the contract won't have enough funds to do so.
```

Comment:
This would always revert btw, because we OpenZeppelin's `ERC721` contract will not allow minting the NFT to the zero address.

https://github.com/RamanSB/Audits/blob/master/puppy-raffle-audit.pdf

See [M-2] & [M-3]

Reply:
Ahh indeed! Thanks for pointing that out :)